### PR TITLE
[MER-1726][Feature] New course creation wizard

### PR DESCRIPTION
**Pull request by augusto1024**
Tickets: [MER-1727](https://eliterate.atlassian.net/browse/MER-1727) and [MER-1728](https://eliterate.atlassian.net/browse/MER-1728)

[MER-1727]: https://eliterate.atlassian.net/browse/MER-1727?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[MER-1728]: https://eliterate.atlassian.net/browse/MER-1728?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: lib/oli_web/live/common/stepper/stepper.ex [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+
+    ~H"""
+      <div id={@id} class="flex md:flex-row flex-col-reverse h-screen w-full">
+        <div class="bg-delivery-body-color-dark dark:bg-gray-900 h-3/5 w-full md:h-full md:w-3/5" />
+        <div class="bg-blue-700 h-2/5 w-full md:h-full md:w-2/5" />
+        <div class="flex md:flex-row flex-col-reverse absolute p-16 top-0 bottom-0 left-0 right-0 m-auto">
+          <div class="bg-white dark:bg-gray-800 w-full md:w-2/3 border border-gray-200 dark:border-gray-600 flex flex-col overflow-hidden">
+            <%= @selected_step.render_fn.(@data) %>
+            <div class="px-9 py-4 flex items-center justify-between border-gray-200 dark:border-gray-600 border-t">
+              <button class="torus-button secondary">Cancel</button>
```
Comment (nicocirio) [inline lib/oli_web/live/common/stepper/stepper.ex:42]: We can add a `phx-click` binding with the corresponding `handle_event` callback to make this button actionable

File: priv/repo/migrations/20230425142146_add_on_delete_to_contained_pages.exs [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+    alter table(:contained_pages) do
+      modify(:section_id, references(:sections, on_delete: :delete_all))
+    end
+  end
+
+  def down do
+    drop(constraint(:contained_pages, "contained_pages_section_id_fkey"))
+
+    alter table(:contained_pages) do
+      modify(:section_id, references(:sections, on_delete: :noting))
```
Comment (nicocirio) [inline priv/repo/migrations/20230425142146_add_on_delete_to_contained_pages.exs:16]: we have a typo here on `nothing`

File: lib/oli_web/live/new_course/select_source.ex [hunk 3]
```diff
@@ -0,0 +1,268 @@
...
+      ) do
+    if !socket.assigns[:loaded] do
+      context = SessionContext.init(session)
+
+      # live_action is :independent_learner, :admin or :lms_instructor
+      live_action = session["live_action"]
+
+      sources =
+        retrieve_all_sources(live_action, %{user: current_user, lti_params: lti_params})
+        |> Enum.with_index(fn element, index -> Map.put(element, :unique_id, index) end)
```
Comment (nicocirio) [inline lib/oli_web/live/new_course/select_source.ex:53]: NIT: maybe this `Enum` could be merged in the previous `retrieve_all_sources` function, to increase readability?

File: lib/oli_web/live/new_course/select_source.ex [hunk 4]
```diff
@@ -0,0 +1,268 @@
...
+  defp mark_selected_row(table_model, nil), do: table_model
+
+  defp mark_selected_row(table_model, source) do
+    source_id =
+      case Regex.run(~r/[^\d]*(\d*)/, source, capture: :all_but_first) |> Enum.at(0) do
+        nil -> nil
+        id -> String.to_integer(id)
+      end
+
+    IO.inspect(source_id)
```
Comment (nicocirio) [inline lib/oli_web/live/new_course/select_source.ex:161]: found some debugging traces here 😄

**Review by nicocirio — COMMENTED** [review]
Great work @augusto1024 !!! the stepper looks **REALLY** good! ❤️

File: lib/oli_web/live/new_course/select_source.ex [hunk 5]
```diff
@@ -0,0 +1,268 @@
...
+  def handle_event("apply_search", _, socket) do
+    filtered_rows = filter_rows(socket.assigns.sources, socket.assigns.params.query)
+
+    table_model = Map.put(socket.assigns.table_model, :rows, filtered_rows)
+
+    {:noreply, assign(socket, table_model: table_model)}
+  end
+
+  def handle_event("reset_search", _, socket) do
+    filtered_rows = filter_rows(socket.assigns.sources, "")
```
[the remark this answers is not in the document]
Comment (augusto1024) [reply lib/oli_web/live/new_course/select_source.ex:211]: Yeah, it definitley should. I just copy the code from the previous `SelectSource` component and I didn't even read it 😅

**Comment by darrensiegel** [comment]
If I try to create a new course section as an Admin from "Manage Open and Free Sections", I see an error:

```
ArgumentError at GET /admin/open_and_free/create
assign @title not available in template.

Please make sure all proper assigns have been set. If this
is a child template, ensure assigns are given explicitly by
the parent template as they are not automatically forwarded.

Available assigns: [:__changed__, :changeset, :conn, :content, :current_author, :current_step, :current_user, :flash, :form_id, :inner_content, :live_action, :live_module, :lti_params, :session, :steps]

[lib/oli_web/templates/layout/_workspace_header.html.eex](https://localhost/admin/open_and_free/create)
<div class="workspace-header-container flex-1 flex flex-column shadow">
  <div class="workspace-header flex-grow-1 flex flex-row flex-nowrap px-6 <%= if Map.has_key?(@conn.assigns, :breadcrumbs) and length(@breadcrumbs) > 0, do: "with-breadcrumbs", else: "" %>">
    <h3><%= if Map.has_key?(@conn.assigns, :breadcrumbs) and length(@breadcrumbs) > 0, do: hd(Enum.reverse(@breadcrumbs)).short_title, else: @title %></h3>
    <div class="flex-1"></div>
    <div>
      <%= render OliWeb.SharedView, "_help_button.html", conn: @conn %>
    </div>
    <div>
OliWeb.LayoutView."_workspace_header.html"/1oli
```

**Comment by augusto1024** [comment]
> If I try to create a new course section as an Admin from "Manage Open and Free Sections", I see an error:
> 
> ```
> ArgumentError at GET /admin/open_and_free/create
> assign @title not available in template.
> 
> Please make sure all proper assigns have been set. If this
> is a child template, ensure assigns are given explicitly by
> the parent template as they are not automatically forwarded.
> 
> Available assigns: [:__changed__, :changeset, :conn, :content, :current_author, :current_step, :current_user, :flash, :form_id, :inner_content, :live_action, :live_module, :lti_params, :session, :steps]
> 
> [lib/oli_web/templates/layout/_workspace_header.html.eex](https://localhost/admin/open_and_free/create)
> <div class="workspace-header-container flex-1 flex flex-column shadow">
>   <div class="workspace-header flex-grow-1 flex flex-row flex-nowrap px-6 <%= if Map.has_key?(@conn.assigns, :breadcrumbs) and length(@breadcrumbs) > 0, do: "with-breadcrumbs", else: "" %>">
>     <h3><%= if Map.has_key?(@conn.assigns, :breadcrumbs) and length(@breadcrumbs) > 0, do: hd(Enum.reverse(@breadcrumbs)).short_title, else: @title %></h3>
>     <div class="flex-1"></div>
>     <div>
>       <%= render OliWeb.SharedView, "_help_button.html", conn: @conn %>
>     </div>
>     <div>
> OliWeb.LayoutView."_workspace_header.html"/1oli
> ```

@darrensiegel Yes, I noticed that while writing the tests. The route for admins uses a workspace layout and I forgot to add the breadcrumbs to the socket in the wizard. I already fixed it in another commit that I'll push later

**Pull request merged**